src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
         this.trit_state = new int[Sponge.HASH_LENGTH];
     }
 
     @Override
     public void reset() {
-
         this.keccak.reset();
     }
 
     @Override
-    public void absorb(final int[] trits, int offset, int length) {
-
+    public void absorb(final int[] trits, final int offset, final int length) {
         if (length % 243 != 0) {
             throw new RuntimeException("Illegal length: " + length);
         }
-
-        do {
+        for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
             //copy trits[offset:offset+length]
-            System.arraycopy(trits, offset, trit_state, 0, HASH_LENGTH);
+            System.arraycopy(trits, pos, trit_state, 0, HASH_LENGTH);
 
             //convert to bits
             trit_state[HASH_LENGTH - 1] = 0;
-            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state, 0);
+            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state);
 
             //run keccak
             keccak.update(byte_state);
-            offset += HASH_LENGTH;
-
-        } while ((length -= HASH_LENGTH) > 0);
+        }
     }
 
     @Override
-    public void squeeze(final int[] trits, int offset, int length) {
-
+    public void squeeze(final int[] trits, final int offset, final int length) {
         if (length % 243 != 0) {
-            throw new RuntimeException("Illegal length: " + length);
+            throw new IllegalArgumentException("Illegal length: " + length);
         }
-
         try {
-          do {
-              this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
-              //convert to trits
-              tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
-
-              //copy with offset
-              trit_state[HASH_LENGTH - 1] = 0;
-              System.arraycopy(trit_state, 0, trits, offset, HASH_LENGTH);
-
-              //calculate hash again
-              for (int i = byte_state.length; i-- > 0; ) {
-
-                  byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
-              }
-              keccak.update(byte_state);
-              offset += HASH_LENGTH;
-
-          } while ((length -= HASH_LENGTH) > 0);
+            for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
+                this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
+                //convert to trits
+                tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
+
+                //copy with offset
+                trit_state[HASH_LENGTH - 1] = 0;
+                System.arraycopy(trit_state, 0, trits, pos, HASH_LENGTH);
+
+                //calculate hash again
+                for (int i = byte_state.length; i-- > 0; ) {
+                    byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
+                }
+                keccak.update(byte_state);
+            }
```

Comment:
Same here - traditional for loop instead of tracking both offset and length. Inputs also final now.